arangod/RestServer/DatabaseFeature.cpp

```diff
@@ -596,9 +597,18 @@ void DatabaseFeature::recoveryDone() {
 
   // '_pendingRecoveryCallbacks' will not change because
   // !StorageEngine.inRecovery()
+  // It's single active thread before recovery done,
+  // so we could use general purpose thread pool for this
+  std::vector<futures::Future<Result>> futures;
+  futures.reserve(_pendingRecoveryCallbacks.size());
   for (auto& entry : _pendingRecoveryCallbacks) {
-    auto result = entry();
-
+    futures.emplace_back(SchedulerFeature::SCHEDULER->queueWithFuture(
+        RequestLane::CLIENT_SLOW, std::move(entry)));
+  }
+  _pendingRecoveryCallbacks.clear();
+  // TODO(MBkkt) use single wait when we start to using normal futures library
+  for (auto& future : futures) {
+    auto result = std::move(future).get();
```

Comment:
There is `futures::collectAll`. Maybe this is what you're looking for. If you want early termination you can add this easily.

Reply:
I know

Reply:
Does it make sense to cancel not-started futures in case of error?

Reply:
Not started should be canceled via arangodb scheduler.
Anyway fast shutdown in case of task not started yet is problem of arangodb scheduler.
It shouldn't resolved locally.